lib/java/com/google/android/material/drawable/DrawableUtils.java

```diff
@@ -207,46 +241,70 @@ public static Drawable compositeTwoLayeredDrawable(
+      topLayerNewWidth = topLayerIntrinsicWidth;
+      topLayerNewHeight = topLayerIntrinsicHeight;
     } else {
-      float topLayerRatio =
-          (float) topLayerDrawable.getIntrinsicWidth() / topLayerDrawable.getIntrinsicHeight();
-      float bottomLayerRatio =
-          (float) bottomLayerDrawable.getIntrinsicWidth()
-              / bottomLayerDrawable.getIntrinsicHeight();
+      topLayerNewWidth = bottomLayerIntrinsicWidth;
+      topLayerNewHeight = bottomLayerIntrinsicHeight;
+    }
+
+    // If both layers have intrinsic size, scale the top layer so that it's not larger than the
+    // bottom one. If the top layer is smaller than the bottom layer in all dimensions,
+    // don't need to do anything.
+    if ((topLayerNewWidth != -1 && topLayerNewHeight != -1)
+        && (bottomLayerIntrinsicWidth != -1 && bottomLayerIntrinsicHeight != -1)
+        && (topLayerNewWidth > bottomLayerIntrinsicWidth
+        || topLayerNewHeight > bottomLayerIntrinsicHeight)) {
+      float topLayerRatio = (float) topLayerNewWidth / topLayerNewHeight;
+      float bottomLayerRatio = (float) bottomLayerIntrinsicWidth / bottomLayerIntrinsicHeight;
       if (topLayerRatio >= bottomLayerRatio) {
         // If the top layer is wider in ratio than the bottom layer, shrink it according to its
         // width.
-        topLayerNewWidth = bottomLayerDrawable.getIntrinsicWidth();
-        topLayerNewHeight = (int) (topLayerNewWidth / topLayerRatio);
+        topLayerNewWidth = bottomLayerIntrinsicWidth;
+        topLayerNewHeight = (int) (bottomLayerIntrinsicWidth / topLayerRatio);
       } else {
         // If the top layer is taller in ratio than the bottom layer, shrink it according to its
         // height.
-        topLayerNewHeight = bottomLayerDrawable.getIntrinsicHeight();
-        topLayerNewWidth = (int) (topLayerRatio * topLayerNewHeight);
+        topLayerNewWidth = (int) (topLayerRatio * bottomLayerIntrinsicHeight);
+        topLayerNewHeight = bottomLayerIntrinsicHeight;
       }
     }
+
     // Centers the top layer inside the bottom layer. Before M there's no layer gravity support, we
     // need to use layer insets to adjust the top layer position manually.
     if (VERSION.SDK_INT >= VERSION_CODES.M) {
-      drawable.setLayerSize(1, topLayerNewWidth, topLayerNewHeight);
+      if (topLayerNewWidth != -1 && topLayerNewHeight != -1) {
+        drawable.setLayerSize(1, topLayerNewWidth, topLayerNewHeight);
+      }
       drawable.setLayerGravity(1, Gravity.CENTER);
     } else {
-      int horizontalInset = (bottomLayerDrawable.getIntrinsicWidth() - topLayerNewWidth) / 2;
-      int verticalInset = (bottomLayerDrawable.getIntrinsicHeight() - topLayerNewHeight) / 2;
-      drawable.setLayerInset(1, horizontalInset, verticalInset, horizontalInset, verticalInset);
+      //noinspection StatementWithEmptyBody
+      if (topLayerNewWidth != -1 && topLayerNewHeight != -1) {
```

Comment:
Please flatten the nested ifs here. Something like:

if (topLayerNewWidth == -1 
    || topLayerNewHeight == -1 
    || bottomLayerIntrinsicWidth == -1
    || bottomLayerIntrinsicHeight == -1) {
  Log.w(...)
} else {
 ...
}
  
}